### ISSUE TITLE
Memory leak in Stetho 1.2.0

### ISSUE THREAD

**Issue by realdeadfish** [facebook/stetho#319]
We've experienced a severe memory leak after we integrated Stetho into our app. Through bisection we could pin down the exact revision when we did this and could verify that - once Stetho was removed - the memory leak was gone.

The leak showed up in Android's Memory Monitor, where constantly around 20KB/s have been added on top until the limit was reached and GC kicked in. Then, from the lower level onwards, memory consumption raised again.

The way we integrated Stetho was fairly basic, we only enabled it in our custom Application instance via `Stetho.initializeWithDefaults(this)` for `DEBUG` builds.

**Comment by jasta** [comment]
What version of Android / device?  What version of Android Studio?  We've had these reports trickle in here and there for almost all of Stetho's releases but I have never been able to reproduce locally.  If you could, please test the Stetho sample app for the leak as well as that is what I have used to reproduce and could not.

I understand why you would not want to send it to us, but if you could take a look at a heap dump locally to see where the possible leak suspects are that would be very helpful as well.  The basic process is available in Android Studio / Android Monitor where you can collect a heap dump and open it in a tool called Eclipse Memory Analyzer.  There is a pretty slick wizard to let you view the potential leak suspects and it should show up pretty easily.

P.S. If you send me your heap file I can do this analysis but it could potentially leak very sensitive information such as auth tokens and what-not.  I do not recommend that you send it unless you are absolutely certain there's nothing sensitive in the file...

**Comment by MohMah** [comment]
I approve of the memory leak, my app usually takes up around 40MB of memory but after I included stetho it gradually increased up to 200MB and then multiple GCs tried to keep it under 200MB. I tried to do a memory dump but due to the low transfer rate and huge memory usage it couldn't download the memory dump from app to my pc.

**Comment by jasta** [comment]
@MohMah, can you help us track this down?  I've tested this extensively on my devices and have not been able to repro.  At least letting me know what Android version and device you have would be really helpful.  If you could try other devices as well (in particular any Nexus series device) that would be helpful too.  If you can confirm it doesn't repro on Nexus devices then we're on to something...

**Comment by ldavin** [comment]
I reproduce the issue mentioned by @tommyd3mdi on my emulator (Google Apis - Android 23 - x86_64 - rev 11) when the following conditions are met:
* App is launched on emulator
* Chrome is launched on my mac, and displays the `chrome://inspect/` page

I've added an allocation tracking file (app is idle so it should only display Stetho allocations/leaks), and a heap dump.
[Allocations-and-snapshot.zip](https://github.com/facebook/stetho/files/78481/Allocations-and-snapshot.zip)

I also reproduce on my Moto X (2nd gen) running on Android 6.0

Thanks for all the work, and happy debugging :)

**Comment by jasta** [comment]
@ldavin thanks so much, I'll take a look shortly and hopefully can figure out what's going on here.

**Comment by KChernenko** [comment]
I'm using `1.3.1` version and also have a memory leak.
![screenshot_2016-03-04-12-44-17](https://cloud.githubusercontent.com/assets/3322440/13525145/ef64d802-e207-11e5-8922-f22b00747c33.png)

**Comment by Heycz** [comment]
@KChernenko I have the same problem

**Comment by jasta** [comment]
Is the only issue that leak canary believes there is a leak here?  I finally analyzed @ldavin 's trace (*embarrassed by how look it took*) and it reveals that the only activity retained by Stetho's mActivities list is indeed also retained by the system itself (ActivityThread$ActivityClientRecord, PhoneWindow$DecorView, etc) which means that there is actually no leak here.  Stetho could, of course, convert this array to a WeakReference which would likely silence LeakCanary's complaint but it seems more so that LeakCanary is actually flagging this as a false positive since Stetho will indeed remove the reference once the system removes its reference.

I am open to the idea of modifying Stetho to use a WeakReference here as it's not a functional change at all and not technically wrong either, but it is worth noting that this seems to be a LeakCanary issue.

**Comment by brianchu** [comment]
I have a similar problem. With the library, my media heavy app runs about few round of video and die. Removing the library solve the problem. The stack when it dies:

06-02 06:23:29.812 W/System.err(11169): java.io.IOException: Too many open files
06-02 06:23:29.812 W/System.err(11169): 	at android.net.LocalSocketImpl.accept(Native Method)
06-02 06:23:29.812 W/System.err(11169): 	at android.net.LocalSocketImpl.accept(LocalSocketImpl.java:328)
06-02 06:23:29.812 W/System.err(11169): 	at android.net.LocalServerSocket.accept(LocalServerSocket.java:94)
06-02 06:23:29.812 W/System.err(11169): 	at com.facebook.stetho.server.LocalSocketServer.listenOnAddress(LocalSocketServer.java:85)
06-02 06:23:29.812 W/System.err(11169): 	at com.facebook.stetho.server.LocalSocketServer.run(LocalSocketServer.java:74)
06-02 06:23:29.812 W/System.err(11169): 	at com.facebook.stetho.server.ServerManager$1.run(ServerManager.java:40)

**Comment by jasta** [comment]
@brianchu that problem is unrelated I believe.  See #308, though the issue is closed because we assumed we had a fix.  If you're still seeing this in the latest release, re-open the issue and we can discuss further.

**Comment by ninovanhooff** [comment]
Still seeing this in 1.4.2. com.facebook.stetho.server.onAccepted() shows a lot of allocations when profiling

**Comment by ZakTaccardi** [comment]
@jasta please convert stetho's reference to a weak one to silence leak canary, would really help out

**Comment by jasta** [comment]
I see now why this is causing an issue.  There's a long list of exclusions in https://github.com/square/leakcanary/blob/master/leakcanary-android/src/main/java/com/squareup/leakcanary/AndroidExcludedRefs.java that excludes the references Android itself keeps to the activity.  So, since Stetho is not in this list, it assumes we're a leak even though we're more a leak suspect.  I agree with the many commenters that the right fix is to use a WeakReference to remove suspicion.

Sorry for dragging my feet on this issue :\

**Comment by ilyaklyukin** [comment]
@jasta  Could you make a new release with this fix? (for Maven rep)

**Comment by jasta** [comment]
Yup!  We're planning a release tentatively this week with a slew of new
features.

>

**Comment by ilyaklyukin** [comment]
@jasta  Any progress?

**Comment by jasta** [comment]
This should be fixed in 1.5.0, which is out now.

**Comment by caipivara** [comment]
@jasta Im having another memory leak on `1.5.0` but Im not sure if I should put it in another issue.

It happens using `'com.facebook.stetho:stetho-okhttp:1.5.0'`

Report:
```
In com.my_app:2.0.0-staging:1.
* com.my_app.screens.login.LoginActivity has leaked:
* GC ROOT com.facebook.stetho.server.ServerManager$1.<Java Local> (anonymous subclass of java.lang.Thread)
* references com.facebook.stetho.server.LocalSocketServer.mSocketHandler
* references com.facebook.stetho.server.LazySocketHandler.mSocketHandlerFactory
* references com.facebook.stetho.Stetho$Initializer$RealSocketHandlerFactory.this$0
* references com.facebook.stetho.Stetho$1.val$context (anonymous subclass of com.facebook.stetho.Stetho$Initializer)
* leaks com.my_app.screens.login.LoginActivity instance

* Retaining: 280 KB.
* Reference Key: 4d2a2c18-232a-4757-9eda-227e659d29aa
* Device: LGE google Nexus 5 hammerhead
* Android Version: 6.0.1 API: 23 LeakCanary: 1.5 00f37f5
* Durations: watch=5015ms, gc=155ms, heap dump=16210ms, analysis=325352ms

* Details:
* Instance of com.facebook.stetho.server.ServerManager$1
| this$0 = com.facebook.stetho.server.ServerManager@319790096 (0x130f9c10)
| val$server = com.facebook.stetho.server.LocalSocketServer@314573392 (0x12c00250)
| contextClassLoader = dalvik.system.PathClassLoader@317471360 (0x12ec3a80)
| daemon = false
| group = java.lang.ThreadGroup@1894715472 (0x70ef1050)
| hasBeenStarted = true
| id = 1811
| inheritableValues = null
| interruptActions = java.util.ArrayList@319936800 (0x1311d920)
| localValues = java.lang.ThreadLocal$Values@319936832 (0x1311d940)
| lock = java.lang.Object@319790080 (0x130f9c00)
| name = java.lang.String@319772928 (0x130f5900)
| nativePeer = -1584569600
| parkBlocker = null
| parkState = 1
| priority = 5
| stackSize = 0
| target = null
| uncaughtHandler = null
| shadow$_klass_ = com.facebook.stetho.server.ServerManager$1
| shadow$_monitor_ = 0
* Instance of com.facebook.stetho.server.LocalSocketServer
| static MAX_BIND_RETRIES = 2
| static $staticOverhead = byte[24]@316576977 (0x12de94d1)
| static WORKER_THREAD_NAME_PREFIX = java.lang.String@314920576 (0x12c54e80)
| static TIME_BETWEEN_BIND_RETRIES_MS = 1000
| mAddress = java.lang.String@314603168 (0x12c076a0)
| mFriendlyName = java.lang.String@1891827184 (0x70c2fdf0)
| mListenerThread = com.facebook.stetho.server.ServerManager$1@314603072 (0x12c07640)
| mServerSocket = android.net.LocalServerSocket@314577168 (0x12c01110)
| mSocketHandler = com.facebook.stetho.server.LazySocketHandler@319789936 (0x130f9b70)
| mStopped = false
| mThreadId = java.util.concurrent.atomic.AtomicInteger@319789952 (0x130f9b80)
| shadow$_klass_ = com.facebook.stetho.server.LocalSocketServer
| shadow$_monitor_ = 0
* Instance of com.facebook.stetho.server.LazySocketHandler
| mSocketHandler = null
| mSocketHandlerFactory = com.facebook.stetho.Stetho$Initializer$RealSocketHandlerFactory@319789968 (0x130f9b90)
| shadow$_klass_ = com.facebook.stetho.server.LazySocketHandler
| shadow$_monitor_ = 0
* Instance of com.facebook.stetho.Stetho$Initializer$RealSocketHandlerFactory
| this$0 = com.facebook.stetho.Stetho$1@319789984 (0x130f9ba0)
| shadow$_klass_ = com.facebook.stetho.Stetho$Initializer$RealSocketHandlerFactory
| shadow$_monitor_ = 0
* Instance of com.facebook.stetho.Stetho$1
| val$context = com.my_app.screens.login.LoginActivity@315602272 (0x12cfb560)
| mContext = com.my_app.base.MyApp@317484192 (0x12ec6ca0)
| shadow$_klass_ = com.facebook.stetho.Stetho$1
| shadow$_monitor_ = 0
* Instance of com.my_app.screens.login.LoginActivity
| static $staticOverhead = byte[16]@316964865 (0x12e48001)
| static serialVersionUID = -5208305141778727029
| static $change = null
| _$_findViewCache = java.util.HashMap@318727792 (0x12ff6670)
| _$_findViewCache = null
| mDebugDrawer = debug_artist.menu.drawer.DebugDrawer@318719040 (0x12ff4440)
| _$_findViewCache = null
| activityComponent$delegate = kotlin.SynchronizedLazyImpl@318752416 (0x12ffc6a0)
| progressWeakReference = java.lang.ref.WeakReference@317782848 (0x12f0fb40)
| mDelegate = android.support.v7.app.AppCompatDelegateImplV23@317687680 (0x12ef8780)
| mEatKeyUpEvent = false
| mResources = null
| mThemeId = 2131361987
| mCreated = true
| mFragments = android.support.v4.app.FragmentController@318717632 (0x12ff3ec0)
| mHandler = android.support.v4.app.FragmentActivity$1@318600064 (0x12fd7380)
| mNextCandidateRequestIndex = 0
| mOptionsMenuInvalidated = false
| mPendingFragmentActivityResults = android.support.v4.util.SparseArrayCompat@318752448 (0x12ffc6c0)
| mReallyStopped = true
| mRequestedPermissionsFromFragment = false
| mResumed = false
| mRetaining = false
| mStopped = true
| mStartedActivityFromFragment = false
| mStartedIntentSenderFromFragment = false
| mExtraDataMap = android.support.v4.util.SimpleArrayMap@318752480 (0x12ffc6e0)
| mActionBar = null
| mActionModeTypeStarting = 0
| mActivityInfo = android.content.pm.ActivityInfo@318677248 (0x12fea100)
| mActivityTransitionState = android.app.ActivityTransitionState@318737024 (0x12ff8a80)
| mApplication = com.my_app.base.MyApp@317484192 (0x12ec6ca0)
| mCalled = true
| mChangeCanvasToTranslucent = false
| mChangingConfigurations = false
| mComponent = android.content.ComponentName@318717648 (0x12ff3ed0)
| mConfigChangeFlags = 0
| mCurrentConfig = android.content.res.Configuration@318741184 (0x12ff9ac0)
| mDecor = null
| mDefaultKeyMode = 0
| mDefaultKeySsb = null
| mDestroyed = true
| mDoReportFullyDrawn = false
| mEmbeddedID = null
| mEnableDefaultActionBarUp = false
| mEnterTransitionListener = android.app.SharedElementCallback$1@1894925104 (0x70f24330)
| mExitTransitionListener = android.app.SharedElementCallback$1@1894925104 (0x70f24330)
| mFinished = true
| mFragments = android.app.FragmentController@318717664 (0x12ff3ee0)
| mHandler = android.os.Handler@318752512 (0x12ffc700)
| mHasCurrentPermissionsRequest = false
| mIdent = 76314953
| mInstanceTracker = android.os.StrictMode$InstanceTracker@318717680 (0x12ff3ef0)
| mInstrumentation = android.app.Instrumentation@318384416 (0x12fa2920)
| mIntent = android.content.Intent@318737088 (0x12ff8ac0)
| mLastNonConfigurationInstances = null
| mMainThread = android.app.ActivityThread@314585344 (0x12c03100)
| mManagedCursors = java.util.ArrayList@318752544 (0x12ffc720)
| mManagedDialogs = null
| mMenuInflater = null
| mParent = null
| mReferrer = java.lang.String@318727840 (0x12ff66a0)
| mResultCode = 0
| mResultData = null
| mResumed = false
| mSearchEvent = null
| mSearchManager = null
| mStartedActivity = false
| mStopped = true
| mTemporaryPause = false
| mTitle = java.lang.String@317527552 (0x12ed1600)
| mTitleColor = 0
| mTitleReady = true
| mToken = android.os.BinderProxy@318550816 (0x12fcb320)
| mTranslucentCallback = null
| mUiThread = java.lang.Thread@1956397728 (0x749c42a0)
| mVisibleBehind = false
| mVisibleFromClient = true
| mVisibleFromServer = true
| mVoiceInteractor = null
| mWindow = com.android.internal.policy.PhoneWindow@315884608 (0x12d40440)
| mWindowAdded = true
| mWindowManager = android.view.WindowManagerImpl@318551968 (0x12fcb7a0)
| mInflater = com.android.internal.policy.PhoneLayoutInflater@318562704 (0x12fce190)
| mOverrideConfiguration = null
| mResources = android.content.res.Resources@317477360 (0x12ec51f0)
| mTheme = android.content.res.Resources$Theme@318552000 (0x12fcb7c0)
| mThemeResource = 2131361987
| mBase = android.app.ContextImpl@318250752 (0x12f81f00)
| shadow$_klass_ = com.my_app.screens.login.LoginActivity
| shadow$_monitor_ = 1238054563
* Excluded Refs:
| Field: android.view.inputmethod.InputMethodManager.mNextServedView
| Field: android.view.inputmethod.InputMethodManager.mServedView
| Field: android.view.inputmethod.InputMethodManager.mServedInputConnection
| Field: android.view.inputmethod.InputMethodManager.mCurRootView
| Field: android.os.UserManager.mContext
| Field: android.net.ConnectivityManager.sInstance
| Field: android.view.Choreographer$FrameDisplayEventReceiver.mMessageQueue (always)
| Thread:FinalizerWatchdogDaemon (always)
| Thread:main (always)
| Thread:LeakCanary-Heap-Dump (always)
| Class:java.lang.ref.WeakReference (always)
| Class:java.lang.ref.SoftReference (always)
| Class:java.lang.ref.PhantomReference (always)
| Class:java.lang.ref.Finalizer (always)
| Class:java.lang.ref.FinalizerReference (always)
In com.my_app:2.0.0-staging:1.
* com.my_app.screens.login.LoginActivity has leaked:
* GC ROOT com.facebook.stetho.server.ServerManager$1.<Java Local> (anonymous subclass of java.lang.Thread)
* references com.facebook.stetho.server.LocalSocketServer.mSocketHandler
* references com.facebook.stetho.server.LazySocketHandler.mSocketHandlerFactory
* references com.facebook.stetho.Stetho$Initializer$RealSocketHandlerFactory.this$0
* references com.facebook.stetho.Stetho$1.val$context (anonymous subclass of com.facebook.stetho.Stetho$Initializer)
* leaks com.my_app.screens.login.LoginActivity instance

* Retaining: 280 KB.
* Reference Key: 4d2a2c18-232a-4757-9eda-227e659d29aa
* Device: LGE google Nexus 5 hammerhead
* Android Version: 6.0.1 API: 23 LeakCanary: 1.5 00f37f5
* Durations: watch=5015ms, gc=155ms, heap dump=16210ms, analysis=325352ms

* Details:
* Instance of com.facebook.stetho.server.ServerManager$1
| this$0 = com.facebook.stetho.server.ServerManager@319790096 (0x130f9c10)
| val$server = com.facebook.stetho.server.LocalSocketServer@314573392 (0x12c00250)
| contextClassLoader = dalvik.system.PathClassLoader@317471360 (0x12ec3a80)
| daemon = false
| group = java.lang.ThreadGroup@1894715472 (0x70ef1050)
| hasBeenStarted = true
| id = 1811
| inheritableValues = null
| interruptActions = java.util.ArrayList@319936800 (0x1311d920)
| localValues = java.lang.ThreadLocal$Values@319936832 (0x1311d940)
| lock = java.lang.Object@319790080 (0x130f9c00)
| name = java.lang.String@319772928 (0x130f5900)
| nativePeer = -1584569600
| parkBlocker = null
| parkState = 1
| priority = 5
| stackSize = 0
| target = null
| uncaughtHandler = null
| shadow$_klass_ = com.facebook.stetho.server.ServerManager$1
| shadow$_monitor_ = 0
* Instance of com.facebook.stetho.server.LocalSocketServer
| static MAX_BIND_RETRIES = 2
| static $staticOverhead = byte[24]@316576977 (0x12de94d1)
| static WORKER_THREAD_NAME_PREFIX = java.lang.String@314920576 (0x12c54e80)
| static TIME_BETWEEN_BIND_RETRIES_MS = 1000
| mAddress = java.lang.String@314603168 (0x12c076a0)
| mFriendlyName = java.lang.String@1891827184 (0x70c2fdf0)
| mListenerThread = com.facebook.stetho.server.ServerManager$1@314603072 (0x12c07640)
| mServerSocket = android.net.LocalServerSocket@314577168 (0x12c01110)
| mSocketHandler = com.facebook.stetho.server.LazySocketHandler@319789936 (0x130f9b70)
| mStopped = false
| mThreadId = java.util.concurrent.atomic.AtomicInteger@319789952 (0x130f9b80)
| shadow$_klass_ = com.facebook.stetho.server.LocalSocketServer
| shadow$_monitor_ = 0
* Instance of com.facebook.stetho.server.LazySocketHandler
| mSocketHandler = null
| mSocketHandlerFactory = com.facebook.stetho.Stetho$Initializer$RealSocketHandlerFactory@319789968 (0x130f9b90)
| shadow$_klass_ = com.facebook.stetho.server.LazySocketHandler
| shadow$_monitor_ = 0
* Instance of com.facebook.stetho.Stetho$Initializer$RealSocketHandlerFactory
| this$0 = com.facebook.stetho.Stetho$1@319789984 (0x130f9ba0)
| shadow$_klass_ = com.facebook.stetho.Stetho$Initializer$RealSocketHandlerFactory
| shadow$_monitor_ = 0
* Instance of com.facebook.stetho.Stetho$1
| val$context = com.my_app.screens.login.LoginActivity@315602272 (0x12cfb560)
| mContext = com.my_app.base.MyApp@317484192 (0x12ec6ca0)
| shadow$_klass_ = com.facebook.stetho.Stetho$1
| shadow$_monitor_ = 0
* Instance of com.my_app.screens.login.LoginActivity
| static $staticOverhead = byte[16]@316964865 (0x12e48001)
| static serialVersionUID = -5208305141778727029
| static $change = null
| _$_findViewCache = java.util.HashMap@318727792 (0x12ff6670)
| _$_findViewCache = null
| mDebugDrawer = debug_artist.menu.drawer.DebugDrawer@318719040 (0x12ff4440)
| _$_findViewCache = null
| activityComponent$delegate = kotlin.SynchronizedLazyImpl@318752416 (0x12ffc6a0)
| progressWeakReference = java.lang.ref.WeakReference@317782848 (0x12f0fb40)
| mDelegate = android.support.v7.app.AppCompatDelegateImplV23@317687680 (0x12ef8780)
| mEatKeyUpEvent = false
| mResources = null
| mThemeId = 2131361987
| mCreated = true
| mFragments = android.support.v4.app.FragmentController@318717632 (0x12ff3ec0)
| mHandler = android.support.v4.app.FragmentActivity$1@318600064 (0x12fd7380)
| mNextCandidateRequestIndex = 0
| mOptionsMenuInvalidated = false
| mPendingFragmentActivityResults = android.support.v4.util.SparseArrayCompat@318752448 (0x12ffc6c0)
| mReallyStopped = true
| mRequestedPermissionsFromFragment = false
| mResumed = false
| mRetaining = false
| mStopped = true
| mStartedActivityFromFragment = false
| mStartedIntentSenderFromFragment = false
| mExtraDataMap = android.support.v4.util.SimpleArrayMap@318752480 (0x12ffc6e0)
| mActionBar = null
| mActionModeTypeStarting = 0
| mActivityInfo = android.content.pm.ActivityInfo@318677248 (0x12fea100)
| mActivityTransitionState = android.app.ActivityTransitionState@318737024 (0x12ff8a80)
| mApplication = com.my_app.base.MyApp@317484192 (0x12ec6ca0)
| mCalled = true
| mChangeCanvasToTranslucent = false
| mChangingConfigurations = false
| mComponent = android.content.ComponentName@318717648 (0x12ff3ed0)
| mConfigChangeFlags = 0
| mCurrentConfig = android.content.res.Configuration@318741184 (0x12ff9ac0)
| mDecor = null
| mDefaultKeyMode = 0
| mDefaultKeySsb = null
| mDestroyed = true
| mDoReportFullyDrawn = false
| mEmbeddedID = null
| mEnableDefaultActionBarUp = false
| mEnterTransitionListener = android.app.SharedElementCallback$1@1894925104 (0x70f24330)
| mExitTransitionListener = android.app.SharedElementCallback$1@1894925104 (0x70f24330)
| mFinished = true
| mFragments = android.app.FragmentController@318717664 (0x12ff3ee0)
| mHandler = android.os.Handler@318752512 (0x12ffc700)
| mHasCurrentPermissionsRequest = false
| mIdent = 76314953
| mInstanceTracker = android.os.StrictMode$InstanceTracker@318717680 (0x12ff3ef0)
| mInstrumentation = android.app.Instrumentation@318384416 (0x12fa2920)
| mIntent = android.content.Intent@318737088 (0x12ff8ac0)
| mLastNonConfigurationInstances = null
| mMainThread = android.app.ActivityThread@314585344 (0x12c03100)
| mManagedCursors = java.util.ArrayList@318752544 (0x12ffc720)
| mManagedDialogs = null
| mMenuInflater = null
| mParent = null
| mReferrer = java.lang.String@318727840 (0x12ff66a0)
| mResultCode = 0
| mResultData = null
| mResumed = false
| mSearchEvent = null
| mSearchManager = null
| mStartedActivity = false
| mStopped = true
| mTemporaryPause = false
| mTitle = java.lang.String@317527552 (0x12ed1600)
| mTitleColor = 0
| mTitleReady = true
| mToken = android.os.BinderProxy@318550816 (0x12fcb320)
| mTranslucentCallback = null
| mUiThread = java.lang.Thread@1956397728 (0x749c42a0)
| mVisibleBehind = false
| mVisibleFromClient = true
| mVisibleFromServer = true
| mVoiceInteractor = null
| mWindow = com.android.internal.policy.PhoneWindow@315884608 (0x12d40440)
| mWindowAdded = true
| mWindowManager = android.view.WindowManagerImpl@318551968 (0x12fcb7a0)
| mInflater = com.android.internal.policy.PhoneLayoutInflater@318562704 (0x12fce190)
| mOverrideConfiguration = null
| mResources = android.content.res.Resources@317477360 (0x12ec51f0)
| mTheme = android.content.res.Resources$Theme@318552000 (0x12fcb7c0)
| mThemeResource = 2131361987
| mBase = android.app.ContextImpl@318250752 (0x12f81f00)
| shadow$_klass_ = com.my_app.screens.login.LoginActivity
| shadow$_monitor_ = 1238054563
* Excluded Refs:
| Field: android.view.inputmethod.InputMethodManager.mNextServedView
| Field: android.view.inputmethod.InputMethodManager.mServedView
| Field: android.view.inputmethod.InputMethodManager.mServedInputConnection
| Field: android.view.inputmethod.InputMethodManager.mCurRootView
| Field: android.os.UserManager.mContext
| Field: android.net.ConnectivityManager.sInstance
| Field: android.view.Choreographer$FrameDisplayEventReceiver.mMessageQueue (always)
| Thread:FinalizerWatchdogDaemon (always)
| Thread:main (always)
| Thread:LeakCanary-Heap-Dump (always)
| Class:java.lang.ref.WeakReference (always)
| Class:java.lang.ref.SoftReference (always)
| Class:java.lang.ref.PhantomReference (always)
| Class:java.lang.ref.Finalizer (always)
| Class:java.lang.ref.FinalizerReference (always)
```
